lib/execution_engine2/db/MongoUtil.py

```diff
@@ -263,6 +264,47 @@ def check_if_already_finished(job_status):
             return True
         return False
 
+    def update_jobs_to_queued(
+        self, job_id_pairs: List[JobIdPair], scheduler_type: str = "condor"
+    ) -> None:
+        """
+        Updates a list of created jobs to queued. Does not work on jobs that already have gone through a status transition
+        :param job_id_pairs: A list of pairs of Job Ids and Scheduler Ids
+        :param scheduler_type: The scheduler this job was queued in, default condor
+        """
+        bulk_operations = []
+        now = time.time()
+        for job_id_pair in job_id_pairs:
+            if None in job_id_pairs:
+                raise InvalidStatusTransitionException
+            bulk_operations.append(
+                UpdateOne(
+                    {
+                        "_id": ObjectId(job_id_pair.job_id),
+                        "status": {
+                            "$in": [Status.created.value, Status.estimating.value]
+                        },
+                    },
+                    {
+                        "$set": {
+                            "status": Status.queued.value,
+                            "queued": now,
+                            "scheduler_id": job_id_pair.scheduler_id,
+                            "scheduler_type": scheduler_type,
+                        }
+                    },
+                )
+            )
+        mongo_collection = self.config["mongo-jobs-collection"]
+        if bulk_operations:
+            with self.pymongo_client(mongo_collection) as pymongo_client:
+                bwr = pymongo_client[self.mongo_database][mongo_collection].bulk_write(
+                    bulk_operations, ordered=False
+                )
+                assert bwr.modified_count == len(job_id_pairs)
```

Comment:
Is this guaranteed to be true if `bulk_write` returns without throwing an error?

Reply:
it is NOT guaranteed true, because if one of the UpdateNow statements doesn't work, e.g. the job is no longer in a state able to be updated to "queued", it should throw an exception and cause these jobs to get cancelled

Reply:
Sorry, I'm confused - it seems like from what you're saying there are 2 scenarios:

* The job returns without an error and the assert is true
* The job throws and exception and the assert statement is skipped

From that is seems to me that the assert has to be true

Reply:
There are 3 scenarios
* The bulk write returns without error, and the assert is true
* The bulk write returns without error, and the number of Updated documents match is less than the number of jobs going in (one of the documents is now in a cancelled state, so it can't update it the document, so it doesn't count towards the updateCount)
* The job throws an exception and the assert statement is skipped

Reply:
Ok, for the second one then that assert is going to get triggered, but that's an expected scenario right? There's nothing wrong with the mod count being < `job_id_pairs`, right?

Reply:
It's an `if` not an `assert` now

Reply:
It's wrong in that
* The wrong number of kafka messages would get sent out for state transitions to Updated, but yea, the damage is done, the jobs are already sent to HTCondor. This will thrown an exception, and we could in theory use this exception  to call a `condor_rm` on these jobs, so they don't take up the user's slots.

Reply:
I mean they should cancel themselves soon enough anyway right

Reply:
So would the right thing to do be just return the IDs of the jobs that updated correctly from MongoUtils and just sent the Kafka messages for those?

Reply:
Yea, we could query mongo again and then send the kafka messages, that would be possible. Uh in theory they will cancel themselves, but then somewhat affects your priority in the queue and your overall score for jobs.

Reply:
But that happens for everyone that cancels jobs, right? The only difference here is that the person cancelled it quickly.

Reply:
No, for the person that cancelled it quickly the cancel jobs endpoint cannot send the cancellation signal to condor, since the cancel_jobs_endpoint doesn't know the condor job id yet until after this update statement is complete.

Reply:
Hum, would the cancel job endpoint throw an error in that case?

So if I start a job normally and then cancel it, it doesn't affect fair share?

Reply:
No, it doesn't throw an error, it just lets the job run and get cancelled when it runs.

Both affect fair share, but the impact is different on having 1 job vs 100 jobs vs 10000 jobs. How much it really matters I am not sure, as in theory they shouldn't run long, but they may end up reserving many cores which could boost the score.  Its possible it doesn't matter that much, but technically the fair share is affected differently.

I see some details at https://htcondor.readthedocs.io/en/latest/admin-manual/user-priorities-negotiation.html?highlight=Real%20User%20Priority%20(RUP)

Reply:
So just to be clear, this line won't throw an error if `job.scheduler_id` is `None`?
https://github.com/kbase/execution_engine2/blob/develop/lib/execution_engine2/sdk/EE2Status.py#L126

Reply:
Regarding fair share: if starting jobs and then cancelling them - either directly via the Condor API or indirectly via the job polling for the canceled state - affect the fair share state, then for the quick cancel case we're talking about a job that will run an extra X seconds, where X is the poll interval. Is that really enough of an issue to worry about here? Or am I missing something?

Reply:
Right, it won't throw an error.
```
>>> class Job:
...     apple = None
...
>>> a = Job()
>>> a.apple
>>> print(a.apple)
None
>>> print(f"{a.apple}.0")
None.0
```
https://github.com/kbase/execution_engine2/blob/3150622a20945cf7a28a77f8356098013a340445/lib/execution_engine2/utils/Condor.py#L280

Doesn't allow for exceptions to bubble up


I'm thinking for small amounts of jobs it shouldn't, but for large amounts, it might. I'm not sure how many seconds it takes for starting up a job, but if its like 15-30 seconds before the jobRunner files are transferred to the node, resources are calculated for HTCondor,  it adds up, especially if you have 100 jobs running for 15-30 seconds taking up an entire large memory node.

Reply:
But considering that sometimes a single job can run for days and then just die, maybe that's not such a big deal.

Reply:
This conversation is getting long but it's important... the level of complexity in starting a job is going to get quite a bit higher with threading.

Ok, so we still have to figure out what to do when a job is canceled before it's updated in mongo (or if the job is in some other terminal state). 

For non-cancel (error, complete, terminated) terminal states, in my opinion it shouldn't throw an error, and should just leave the jobs as is.

For the canceled jobs, I can think of two choices currently:
1) Leave the jobs alone and let the job poll cancellation method do its thing. This might cost the user a few seconds or minutes of fair share time per job.
2) Maybe do a two stage update:
- update the scheduler ID for all the jobs to make sure it's recorded in the DB regardless of the state.
  - this might be worth doing regardless.
- update jobs that are in the created state to updated.
  - if `n_updated > len(jobs)` then pull the jobs, and return the jobs that aren't in the updated or created state.
  - EE2RunJob calls `cancel_job()` on jobs that are in the canceled state.

Any other ideas?

Reply:
Other idea is that since we know the scheduler ids, we could catch the exception and then cancel the jobs in htcondor with the api.

Reply:
Saving the scheduler ID also aids with forensics in the future

Reply:
>   catch the exception and then cancel the jobs

You mean cancel all the jobs? If so that seems odd to me - if I cancel a single job quickly all my jobs get cancelled, but if I wait a little bit then just the one job gets canceled?

Reply:
Yes cancel all the jobs in HTCondor. Why does that seem odd, you then don't have to worry about having the job start up before it realizes it needs to cancel itself.

Reply:
It seems odd because I told the UI to cancel one of my ten jobs, but they all wind up getting canceled, assuming I'm understanding you.

Reply:
Oooh I see.. I didn't think about being able to cancel one of the jobs, I was thinking of cancelling all of the jobs at once. That would be odd..

In that case, then maybe the way to proceed would be to see if the operation to update all jobs didn't equal the number of jobs, and then go through and do the 2nd update
Update * where job id in [ list of ids we just had] and job.scheduler_id is blank, update the scheduler_id (and set to cancel state just to be sure?) and then call HTCondor.cancel on the scheduler_id....

Actually I just realized you can call an HTCondor.cancel using the ee2 id, without knowing the scheduler_id
condor_rm -constraint 'JobBatchName=="ee2_id"'

Reply:
Ok, so maybe update cancel_job to use the ee2 id instead of the scheduler ID, and then this is moot?

Reply:
Personally I'd still always add the scheduler ID to the mongo record as well because why not

Reply:
Changed it around, removed the exception. 
I was thinking we could also just mark the record as cancelled, but throw an exception telling the user to wait a few seconds and try to cancel again ?

add
      ```  if job.scheduler_id is None:
            raise CondorJobNotFoundException(f"Couldn't cancel {job_id}. Please try again in a few seconds")
```
to the end of cancel_job

Reply:
I don't think you pushed the changes?

> I was thinking we could also just mark the record as cancelled, but throw an exception telling the user to wait a few seconds and try to cancel again ?

But then if the front end refreshes the job will show as cancelled, so why would use user try again (even if the front end lets them, which I assume it won't)

Reply:
Ok pushed. 

I believe in the frontend for batch upload, it doesn't show as cancelled, and it handles cancel_job failures by telling you that the cancellation failed.
Now that I think about it, not actually sure what happens to the cancel job button or the entire cell for single jobs when the cancel endpoint raises an exception.

Reply:
> I believe in the frontend for batch upload, it doesn't show as cancelled,

I don't understand this. If it pulls the job from the backend and the job is in the cancelled state, the front end will ignore that?

Reply:
That would depends on whether or not the next cell state transition depends on what happens after the cancel job is called. If that state transition depends on the successful cancel job api call, it wouldn't transition.

Reply:
Something is in the works right now https://github.com/kbase/narrative/pull/2384

Reply:
I'm talking about the response to the call though, not about the polls. The typical place for the user to cancel a job is in the cell or the job browser, so that's where they want to know about the cancellation failure.

Reply:
We could just add another lookup to see if any of these jobs are now in a terminated state and cancel them if required.

Reply:
I'm not sure we're understanding each other. If I understand you correctly, what'll happen on a fast cancel is:

* The user starts a batch job and cancels at least one of the jobs quickly enough that condor submission hasn't occurred (which could take seconds to minutes depending on the number of jobs)
* The job is marked as canceled, but not actually canceled
* The user gets a message saying to try and cancel again in a few minutes
* But if the user, say, walks away for a bit and then loads to the job broswer to cancel the job, it'll show as already canceled when it actually isn't.

That seems incorrect to me. Please correct me if I'm wrong about what happens

Reply:
Yes that is mostly correct. The job is marked as cancelled, but that is a "partial" cancel and that prevents the jobRunner from starting and ends the job if it's running once the cancel checker runs. If condor is unavailable, not responding there's not much we can do right now about stopping the job in condor once it comes back without a other reaper or Kafka. The jobRunner would have  to query the job state and make a decision right now.  If condor is available, then we could have this thread check for cancellation status right before it ends, and issue out cancel job commands once more.

Reply:
Offline we decided that if a user hits cancel before the jobs are submitted to condor:

* the job will be marked as canceled
* the user won't get any abnormal error message, to them it'll look like the job is canceled
* at the end of the job submission thread, the thread will re-pull the jobs from mongo and send cancel messages to Condor for any cancelled jobs, ignoring any errors that may occur due to the job already being cancelled

Worst case this means that the job runs for, at most, the length of the JobRunner cancellation poll interval before stopping.

@bio-boris if this isn't your understanding let's talk again, but I think this is what we decided.

Reply:
:+1: